lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp

```diff
@@ -0,0 +1,893 @@
+  SmallVectorImpl<size_t>::iterator portEnd;
+
+  void skipOutputsWithNoChild() {
+    auto hasNoChild = [&](Value v) {
+      return ChildIterator(v) == ChildIterator(v, /*end=*/true);
+    };
+
+    while (portIt != portEnd && hasNoChild(instance.getResult(*portIt)))
+      ++portIt;
+    child = portIt == portEnd ? ChildIterator()
+                              : ChildIterator(instance.getResult(*portIt));
+  }
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Subfield child iterator
+//===----------------------------------------------------------------------===//
+
+namespace {
+class SubfieldNodeIterator : public NodeIteratorBase {
+public:
+  explicit SubfieldNodeIterator(SubfieldOp subfield, Node node,
+                                bool end = false)
+      : NodeIteratorBase(node), subfield(subfield) {
+    assert(subfield == node.value.getDefiningOp<SubfieldOp>() &&
+           "subfield must be the defining op of the node value");
+
+    auto memory = subfield.input().getDefiningOp<MemOp>();
+    if (!memory) {
+      subfield->emitOpError("input must be a port of a MemOp, please run "
+                            "-firrtl-lower-types first");
+      return;
+    }
+
+    auto portKind =
+        memory.getPortKind(subfield.input().cast<OpResult>().getResultNumber());
+    if (subfield.fieldIndex() != 0 || memory.readLatency() != 0 ||
+        portKind == MemOp::PortKind::Write)
+      return;
+
+    if (!end) {
+      auto users = subfield.input().getUsers();
+      for (auto it = users.begin(), e = users.end(); it != e; ++it) {
+        auto currentSubfield = dyn_cast<SubfieldOp>(*it);
+        if (!currentSubfield) {
+          it->emitOpError("MemOp must be used by SubfieldOp, please run "
+                          "-firrtl-lower-types first");
+          return;
+        }
+
+        if ((currentSubfield.fieldIndex() == 3 &&
+             portKind == MemOp::PortKind::Read) ||
+            (currentSubfield.fieldIndex() == 4 &&
+             portKind == MemOp::PortKind::ReadWrite)) {
+          child = ChildIterator(currentSubfield.result());
+          dataSubfield = currentSubfield;
+          return;
+        }
+      }
+    }
+  }
+
+  SubfieldNodeIterator &operator++() {
+    auto childEnd = ChildIterator(dataSubfield.result(), /*end=*/true);
+    if (child != childEnd)
+      ++child;
+    if (child == childEnd)
+      child = ChildIterator();
+    return *this;
+  }
+
+private:
+  SubfieldOp subfield;
+  SubfieldOp dataSubfield;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// CombIterator class
+//===----------------------------------------------------------------------===//
+
+using CombIteratorBase = std::variant<DefaultNodeIterator, InstanceNodeIterator,
+                                      SubfieldNodeIterator>;
+
+namespace {
+class CombIterator : public CombIteratorBase {
```

Comment:
I'd recommend containing the variant as a member instead of inheriting from it.  There is no "isa" relationship with the variant type itself, though I get what you're conceptually going for here.